### ISSUE TITLE
Check if `NotifyCommit` messages in TLA+ are safe

### ISSUE THREAD

**Issue by heidihoward** [microsoft/CCF#5280]
We should remove `NotifyCommit` messages from the TLA+ spec as they do not exist in the implementation (`AppendEntries` are used instead). 

However, whilst looking at the spec, I noticed that the `NotifyCommit` messages would be unsafe, however, this hasn't been flag by the model checker. It might be worth looking into why before removing them.

**Comment by lemmy** [comment]
Collecting coverage from simulating the `ccfraft.tla` via `SIMccraft.tla` for 90 minutes shows very low coverage of `NotifyCommit`.   I believe this is because `SIMccfraft.tla` "favors" long-lived leaders s.t. a leader rarely retire.   Consequently, the action `UpdateCommitIndex` has also very low coverage.

<img width="1150" alt="Screenshot 2023-05-18 at 1 18 21 PM" src="https://github.com/microsoft/CCF/assets/88777/9e592a83-b3e5-43c6-b26b-9b5b0f9f598d">
<img width="1073" alt="Screenshot 2023-05-18 at 1 18 58 PM" src="https://github.com/microsoft/CCF/assets/88777/8568ba8f-cdb1-4af9-8171-0a84696a14ba">


It seems `NotifyCommit` ends up being safe because the `commitIndex` is checked at the receiver in `UpdateCommitIndex`:

https://github.com/microsoft/CCF/blob/80274ca82e2faa9156ec5939f6a716ab6e863792/tla/ccfraft.tla#L1034

**Comment by lemmy** [comment]
Hypothesis:

1. A RetiredLeader's `commitIndex` no longer changes because a RetiredLeader rejects `AppendEntriesRequest` https://github.com/microsoft/CCF/blob/80274ca82e2faa9156ec5939f6a716ab6e863792/tla/ccfraft.tla#L961-L964
2. A Leader transitions to RetiredLeader iff the cluster's `commitIndex` is advanced and the Leader is dropped from the configuration.  Thus, at least one majority of the cluster are at the given `commitIndex`
3. A retiring Leader trims the prefix of its configuration up to its `commitIndex` https://github.com/microsoft/CCF/blob/80274ca82e2faa9156ec5939f6a716ab6e863792/tla/ccfraft.tla#L754
4. A RetiredLeader will only send `NotifyCommitMessage` to nodes that are members of the one `configuration` less than or equal `commitIndex` (1. and 3.) https://github.com/microsoft/CCF/blob/80274ca82e2faa9156ec5939f6a716ab6e863792/tla/ccfraft.tla#L783-L787
5. a) The majority nodes' `commitIndex` will be equal (and eventually greater) than the RetiredLeader's `commitIndex` (1.)
    b) Other node's will reject `NotifyCommitMessage` for as long as their `commitIndex` is less than the RetiredLeaders https://github.com/microsoft/CCF/blob/80274ca82e2faa9156ec5939f6a716ab6e863792/tla/ccfraft.tla#L1033-L1034
6. See 5a. once the other nodes catch up

In short, `NotifyMessage` & `UpdateCommitIndex` are one elaborate (algorithmic) no-op.

----

Bonus properties:

```tla
RetiredCommitIndexProp ==
    [][\A s \in Servers: state[s] = RetiredLeader => commitIndex[s] = commitIndex'[s]]_vars

\* A RetiredLeader has exactly one configuration for its commitIndex.
RetiredSingleConfigInv ==
    \A s \in Servers: state[s] = RetiredLeader => Cardinality({d \in DOMAIN configurations[s] : d <= commitIndex[s]}) = 1
```

----

If the above reasoning checks out and `NotifyMessage` & `UpdateCommitIndex` are removed from ccfraft.tla, the purpose of a RetiredLeaders should be made more explicit.  Right now, a RetiredLeader appears useless due to 1.

**Comment by lemmy** [comment]
Follow-up: Why is a RetiredLeader allowed to cast votes in subsequent leader elections?  Can't this lead to liveness issues such as some set of RetiredLeaders repeatedly electing some leader that then cannot make forward progress.

Note that violations of the properties below confirmed that a RetiredLeader can grant its vote in a subsequent leader election (I confirmed that the node first retired and later sent the RequestVoteResponse):

```tla
RetiredLeaderNeverVotesFor ==
    \A m \in {m \in Messages: m.type = RequestVoteResponse}:
        state[m.source] = RetiredLeader =>
            m.voteGranted = FALSE
```
